### PR TITLE
Only try to modify file times of a writable file on Windows

### PR DESCRIPTION
**Pull request by jieyouxu**
Introduces a `set_file_times` helper which opens a given path as a file in r+w mode on Windows and then sets file times. Previously the file was open as read-only for Windows which caused permission errors locally.

This should hopefully make setting file times less error prone, since trying to set file times on read-only file on Windows also happened in #127850.

try-job: dist-loongarch64-musl
try-job: x86_64-msvc

### CODE REVIEW

File: src/bootstrap/src/lib.rs [hunk 1]
```diff
@@ -1777,9 +1779,8 @@ Executed at: {executed_at}"#,
             }
         }
         if let Ok(()) = fs::hard_link(&src, dst) {
-            // Attempt to "easy copy" by creating a hard link
-            // (symlinks don't work on windows), but if that fails
-            // just fall back to a slow `copy` operation.
+            // Attempt to "easy copy" by creating a hard link (symlinks are priviledged on windows),
+            // but if that fails just fall back to a slow `copy` operation.
```
Comment (jieyouxu) [inline src/bootstrap/src/lib.rs:1783]: This was a drive-by comment update because symlinks are priviledge on Windows, but AFAIK they do work when you have the priviledge (but can be weird).

**Comment by Kobzol** [comment]
Documented [here](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-setfiletime) in the Windows API.

LGTM, feel free to r=me once CI is green.

**Comment by matthiaskrgr** [comment]
@bors r-
 
 2024-08-12T08:09:07.8523294Z [1m[32m   Compiling[0m clap_derive v4.5.5
2024-08-12T08:09:08.7043836Z [RUSTC-TIMING] tracing_attributes test:false 1.637
2024-08-12T08:09:08.7050616Z [1m[32m   Compiling[0m thiserror-impl v1.0.61
2024-08-12T08:09:10.1303074Z [RUSTC-TIMING] thiserror_impl test:false 1.419
2024-08-12T08:09:10.1792105Z [RUSTC-TIMING] thiserror test:false 0.043
2024-08-12T08:09:10.1798434Z [1m[32m   Compiling[0m tracing v0.1.37
2024-08-12T08:09:10.2561947Z [RUSTC-TIMING] clap_derive test:false 2.397
2024-08-12T08:09:10.2568760Z [1m[32m   Compiling[0m clap v4.5.7
2024-08-12T08:09:10.2983590Z [RUSTC-TIMING] clap test:false 0.034
2024-08-12T08:09:10.5155182Z [RUSTC-TIMING] tracing test:false 0.330
2024-08-12T08:09:10.5173792Z [1m[32m   Compiling[0m llvm-bitcode-linker v0.0.1 (/checkout/src/tools/llvm-bitcode-linker)
2024-08-12T08:09:11.0202785Z [RUSTC-TIMING] llvm_bitcode_linker test:false 0.498
2024-08-12T08:09:11.1082040Z [RUSTC-TIMING] clap_builder test:false 8.264
2024-08-12T08:09:12.3069458Z [RUSTC-TIMING] llvm_bitcode_linker test:false 1.194
2024-08-12T08:09:12.3083923Z [1m[32m    Finished[0m `release` profile [optimized] target(s) in 11.09s
2024-08-12T08:09:12.3229000Z ##[endgroup]
2024-08-12T08:09:12.3230751Z [TIMING] core::build_steps::tool::LlvmBitcodeLinker { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu }, target: x86_64-unknown-linux-gnu, extra_features: [] } -- 11.115
2024-08-12T08:09:12.3233162Z [TIMING] core::build_steps::compile::Assemble { target_compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu } } -- 0.003
2024-08-12T08:09:12.3234606Z [TIMING] core::build_steps::compile::StartupObjects { compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu }, target: loongarch64-unknown-linux-musl } -- 0.000
2024-08-12T08:09:12.3236187Z [TIMING] core::builder::Builder::sysroot_libdir::Libdir { compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu }, target: loongarch64-unknown-linux-musl } -- 0.000
2024-08-12T08:09:12.3278104Z thread 'main' panicked at src/lib.rs:1794:13:
2024-08-12T08:09:12.3279076Z set_file_times(dst, file_times) failed with Permission denied (os error 13)
2024-08-12T08:09:12.3279792Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2024-08-12T08:09:12.3301323Z Build completed unsuccessfully in 0:16:13
2024-08-12T08:09:12.3381725Z   local time: Mon Aug 12 08:09:12 UTC 2024
2024-08-12T08:09:12.4135264Z   network time: Mon, 12 Aug 2024 08:09:12 GMT
2024-08-12T08:09:12.6295762Z ##[error]Process completed with exit code 1.
2024-08-12T08:09:12.6402805Z Post job cleanup.

**Comment by jieyouxu** [comment]
... wat

**Comment by Kobzol** [comment]
It's possible that some files are not writable on Linux (not sure if it's because of Docker or something else), but it was possible to set their modification times (?). Would be useful to see the path for which the mtime modification failed.

**Comment by ChrisDenton** [comment]
The code that failed is this section: https://github.com/rust-lang/rust/blob/13c36f131b0652658b0daaadde674b8d9f708dbc/src/bootstrap/src/lib.rs#L1785-L1794

It should be writeable because the file was just copied. But setting the permissions may make it unable to be reopen with write access.

**Comment by Kobzol** [comment]
> The code that failed is this section:
> 
> https://github.com/rust-lang/rust/blob/13c36f131b0652658b0daaadde674b8d9f708dbc/src/bootstrap/src/lib.rs#L1785-L1794
> 
> It should be writeable because the file was just copied. But setting the permissions may make it unable to be reopen with write access.

That's a good point. Maybe just reordering the file time modification and the permissions could fix it?

**Comment by jieyouxu** [comment]
> That's a good point. Maybe just reordering the file time modification and the permissions setting could fix it?

(We can't do that AFAICT because as I understand it the set file times is there because `chmod` changes file times, not sure about windows)

File: src/bootstrap/src/lib.rs [hunk 2]
```diff
@@ -1790,8 +1791,7 @@ Executed at: {executed_at}"#,
                 .set_accessed(t!(metadata.accessed()))
                 .set_modified(t!(metadata.modified()));
 
-            let dst_file = t!(fs::File::open(dst));
-            t!(dst_file.set_times(file_times));
+            t!(set_file_times(dst, file_times));
```
Comment (jieyouxu) [inline src/bootstrap/src/lib.rs:1794]: (Locally I added a comment for why this is even needed, due to the set file permissions changing file access times at least on linux).

**Comment by ChrisDenton** [comment]
We could just not reopen the file and instead use the same fd for setting both the permissions and file times.

**Comment by jieyouxu** [comment]
```
could not open /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/loongarch64-unknown-linux-musl/lib/self-contained/libc.a as writable file
```

**Comment by Kobzol** [comment]
Ok, a special case for Windows seems reasonable. Let's try again.

@bors r+ rollup=iffy

**Comment by jieyouxu** [comment]
Fixed PR title and description to match what the changes actually do.

**Comment by matthiaskrgr** [comment]
failed in https://github.com/rust-lang/rust/pull/129129#issuecomment-2291905329 I think

**Comment by workingjubilee** [comment]
almost certainly failed in https://github.com/rust-lang/rust/pull/129142

**Comment by jieyouxu** [comment]
Ah there's a new `Flags` wrapper struct

**Comment by Kobzol** [comment]
It's not new, it's just now required to pass it to `Config`, instead of the raw args. The `Config` struct will receive further refactorings in the near future =D

Thanks!

@bors r+

**Comment by jieyouxu** [comment]
> The `Config` struct will receive further refactorings in the near future =D

Looking forward to it 👀
